### PR TITLE
feat(datetime): add header text to multiple selection; improve header consistency between modes

### PR DESCRIPTION
**Pull request by amandaejohnston**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

- When `multiple="true"` and the header is enabled (e.g. when `showDefaultTitle="true"`), the part of the datetime's header that shows what date you have selected (hereafter referred to as just "the header") is not rendered.
- The header doesn't render in `ios` mode at all, whether `multiple` is enabled or not.

Issue URL: resolves https://github.com/ionic-team/ionic-framework/issues/25668

| `md` mode | `ios` mode |
| ------- | ------- |
| md mode | ios mode |

## What is the new behavior?

- The header now shows if `multiple="true"`. If 0 or more than 1 dates are selected, the header shows the number, rather than the exact selected dates. (The behavior when exactly 1 date is selected is unchanged.)
- Added a new prop `titleSelectedDatesFormatter` which can be used to customize the display of the header for localization purposes.
  - Note that this function is not called when exactly 1 date is selected, since that part is already localized automatically. Otherwise, devs would need to replicate the existing localization in their callback to handle this case.
- The selected date(s) will also now show in the header in `ios` mode. The text is identical between the two modes in all cases.
  - This also applies to single-selection datetimes.

| `md` mode | `ios` mode |
| ------- | ------- |
| md mode | ios mode |

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: core/src/components/datetime/datetime.ios.scss [hunk 1]
```diff
@@ -20,12 +20,16 @@
...
 }
 
 :host .datetime-header .datetime-title {
   color: var(--title-color);
+}
 
-  font-size: 14px;
+:host .datetime-header .datetime-selected-date {
+  @include margin(10px, null, null, null);
 }
```
Comment (amandaejohnston) [inline core/src/components/datetime/datetime.ios.scss:33]: This styling was fairly arbitrary on my part; open to suggestions. I just re-used the same font size and picked a margin that looked good with the existing design.

File: core/src/components/datetime/datetime.tsx [hunk 2]
```diff
@@ -313,6 +314,14 @@ export class Datetime implements ComponentInterface {
    */
   @Prop() firstDayOfWeek = 0;
 
+  /**
+   * A callback used to format the header text that shows how many
+   * dates are selected. Only used if there are 0 or more than 1
+   * selected (i.e. unused for exactly 1). By default, the header
+   * text is set to "datesSelected days".
+   */
+  @Prop() selectedDatesLabelFormatter?: SelectedDatesLabelFormatter;
```
Comment (amandaejohnston) [inline core/src/components/datetime/datetime.tsx:323]: I'm not married to this name if anyone wants to suggest alternatives.

Comment (sean-perkins) [reply core/src/components/datetime/datetime.tsx:323]: Perhaps `headerTextFormatter` or `headerSelectedDatesFormatter`?

File: core/src/components/datetime/datetime-interface.ts [hunk 3]
```diff
@@ -23,3 +23,5 @@ export interface DatetimeParts {
 }
 
 export type DatetimePresentation = 'date-time' | 'time-date' | 'date' | 'time' | 'month' | 'year' | 'month-year';
+
+export type SelectedDatesLabelFormatter = (datesSelected: number) => string;
```
Comment (sean-perkins) [inline core/src/components/datetime/datetime-interface.ts:27]: ~ Should we tweak the naming of `dateSelected` to represent a number? e.g.: `selectedDateCount`, `numberOfDates`, `count`, etc. 

With the current naming it may be mistaken as the array of dates, until they interact with the type.

File: core/src/components/datetime/datetime.tsx [hunk 4]
```diff
@@ -2071,20 +2079,39 @@ export class Datetime implements ComponentInterface {
...
+  private renderCalendarViewHeader() {
     const hasSlottedTitle = this.el.querySelector('[slot="title"]') !== null;
     if (!hasSlottedTitle && !this.showDefaultTitle) {
       return;
     }
 
+    const { activeParts, selectedDatesLabelFormatter } = this;
+    const isArray = Array.isArray(activeParts);
+
+    let headerText;
```
Comment (sean-perkins) [inline core/src/components/datetime/datetime.tsx:2091]:
```suggestion
    let headerText: string;
```

File: core/src/components/datetime/datetime.tsx [hunk 5]
```diff
@@ -2071,20 +2079,39 @@ export class Datetime implements ComponentInterface {
...
+    if (isArray && activeParts.length !== 1) {
+      const defaultText = `${activeParts.length} days`;
```
Comment (sean-perkins) [inline core/src/components/datetime/datetime.tsx:2093]: Can we just the `headerText` here, so we don't need to have additional assignments for exceptions/else checks?

e.g.:
```ts
headerText = `${activeParts.length} days`;
if (selectedDatesLabelFormatter !== undefined) {
  try {
    headerText = selectedDatesLabelFormatter(activeParts.length);
  } catch (e) {
    printIonError('Exception in provided `selectedDatesLabelFormatter`: ', e);
  }
}
```

File: core/src/components/datetime/datetime.tsx [hunk 6]
```diff
@@ -2071,20 +2079,39 @@ export class Datetime implements ComponentInterface {
...
+
     return (
       <div class="datetime-header">
         <div class="datetime-title">
           <slot name="title">Select Date</slot>
         </div>
-        {mode === 'md' && !this.multiple && (
-          <div class="datetime-selected-date">{getMonthAndDay(this.locale, this.activeParts as DatetimeParts)}</div>
-        )}
+        <div class="datetime-selected-date">{headerText}</div>
```
Comment (sean-perkins) [inline core/src/components/datetime/datetime.tsx:2114]: I didn't think of this aspect during the design phase, but should we allow `selectedDatesLabelFormatter` to return undefined to disable the rendering of this space?

Then this area could be conditionally rendered:
```jsx
{headerText && (<div class="datetime-selected-date">{headerText}</div>}
```

**Review by sean-perkins — COMMENTED** [review]
Since this feature changes the behavior of single selection datetime, should we also have similar checks as the multiple tests, to verify that text is rendering correctly (outside of screenshot tests)?

File: core/src/components/datetime/datetime.tsx (hunk 2 above)
Comment (amandaejohnston) [reply core/src/components/datetime/datetime.tsx:323]: Hmm, maybe `titleSelectedDatesFormatter` to match up with the `showDefaultTitle` prop?

**Comment by amandaejohnston** [comment]
> Since this feature changes the behavior of single selection datetime, should we also have similar checks as the multiple tests, to verify that text is rendering correctly (outside of screenshot tests)?

I would say the screenshot tests cover that, since the header text can only show the one selected date -- unlike when `multiple="true"`, where there are a few different cases to account for. I could be convinced otherwise though.

File: core/src/components/datetime/datetime.tsx (hunk 6 above)
Comment (amandaejohnston) [reply core/src/components/datetime/datetime.tsx:2114]: I don't see why not 👀

File: core/src/components/datetime/datetime.tsx (hunk 2 above)
Comment (sean-perkins) [reply core/src/components/datetime/datetime.tsx:326]: That works for me 👍 I like aligning context with `showDefaultTitle`.

**Review by liamdebeasi — COMMENTED** [review]
Is integrating `datetime-button` with these changes part of this work, or would you recommend we do separate design work for that?

File: core/src/components/datetime/datetime.tsx [hunk 7]
```diff
@@ -2071,20 +2082,36 @@ export class Datetime implements ComponentInterface {
...
     ];
   }
-  private renderCalendarViewHeader(mode: Mode) {
+  private renderCalendarViewHeader() {
     const hasSlottedTitle = this.el.querySelector('[slot="title"]') !== null;
     if (!hasSlottedTitle && !this.showDefaultTitle) {
       return;
     }
 
+    const { activeParts, titleSelectedDatesFormatter: titleSelectedDatesFormatter } = this;
```
[the remark this answers is not in the document]
Comment (amandaejohnston) [reply core/src/components/datetime/datetime.tsx]: This is outdated; I fixed that in https://github.com/ionic-team/ionic-framework/pull/25817/commits/7d2c6a8e76ebd5d0179564dcd7d4e7d998e8a042. No idea how it got there 😆

**Comment by amandaejohnston** [comment]
I figure updating `datetime-button` is a separate task.

**Review by liamdebeasi — APPROVED** [review]
Great job! The property name is a bit wordy, but I also can't think of a shorter name that is still descriptive.
